### PR TITLE
[fix] #3247: Removed fault injection from sumeragi.

### PR DESCRIPTION
**Pull request by SamHSmith**

### CODE REVIEW

File: core/benches/kura.rs
Comment (appetrosyan) [reply core/benches/kura.rs:54]: In python we'd use named parameters (`**kwargs`). Here, the next best thing is the struct literal syntax. 

```rust
PendingBlockStub {
   transactions: Vec<VersionedAcceptedTransaction>,
   event_recommendations: Vec<Event>,
   height: u64,
   previous_block_hash: Option<HashOf<VersionedCommittedBlock>>,
   view_change_index: u64,
   committed_with_topology: Topology,
   key_pair: KeyPair,
   transaction_validator: &TransactionValidator,
   wsv: &WorldStateView,
}
```
Where you get the benefit of not worrying about filling in all the details thanks to the ability to only define the struct fields that matter and using `..PendingBlockStub::default()` later. This allows you not to have to specify things like  height (mostly it's `1` unless otherwise specified) and `view_change_index`. 

Finally, you just `impl From<PendingBlockStub> for PendingBlock` and postfix the call to the struct literal with `.into()`.

**Review by appetrosyan — CHANGES_REQUESTED** [review]

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:49]: This function for sure is not a good `inline` candidate, given the register pressure.

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:79]: Consider making this function accept a `WorldStateView` and cloning one level above.

File: core/benches/kura.rs
Comment (appetrosyan) [reply core/benches/kura.rs:54]: After looking at the `new` method, it might be a good idea to just add 
```rust
impl PendingBlockStub { 
  fn build() -> PendingBlock {
  ```

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:97]: Good

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:219]:
```suggestion

```
It's allowed project-wide.

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:248]: Nice to see traits.

Comment (appetrosyan) [reply core/src/block.rs:248]:
```suggestion
pub trait Revalidate: Sized {
```
Rust has a different naming convention for interfaces.

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:302]: Another implicit `wsv` clone.

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:250]: Also, there's a question of whether or not we want for it to be Sized. Revalidate implies that you're using the object Read-only. So why take ownership?

Is it for the chain syntax? 

If there's a good reason for it, there ought to be a comment in the text of the function declaration. But I'd also consider rewriting this with a `&self`.

File: core/src/block.rs
Comment (appetrosyan) [inline core/src/block.rs:626]: Good use of functional style

File: core/src/sumeragi/main_loop.rs
Comment (appetrosyan) [inline core/src/sumeragi/main_loop.rs:69]: Good. I don't mind a manual impl in this case.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:49]: I agree. This was a copy/paste/refactor mistake.

File: core/src/sumeragi/mod.rs
Comment (0x009922) [inline core/src/sumeragi/mod.rs:28]: I see that in other cases module-level doc tests are *usually* located inside of a module, not outside of it.

File: core/src/block.rs
Comment (ilchu) [inline core/src/block.rs:604]: Do you think having a more structured error (e.g. its own error type) in `block` would be beneficial? Getting rid of `eyre` in this module would also allow to get rid of `Sized` bound on `Revalidate`, though I agree it's minor.

File: core/src/block.rs
Comment (ilchu) [inline core/src/block.rs:429]: I'm a bit on the fence about implementing this for `VersionedCommittedBlock` instead of just `CommittedBlock`. If we moved all the helper methods from VCB into CB, both methods in this trait could have a default impl, as the destructuring from `V1` seems to be the only difference. And also semantically it seems to make more sense, as I believe versioning is more of a data model concern, and we shouldn't be dealing with it in core.

File: core/src/sumeragi/main_loop.rs
Comment (ilchu) [inline core/src/sumeragi/main_loop.rs:17]: Great to know that it could be deleted that easily. Non-generic signatures are much easier to read.

**Review by ilchu — CHANGES_REQUESTED** [review]

File: core/src/block.rs
Comment (mversic) [reply core/src/block.rs:604]: `eyre` should be removed from `iroha_core` and errors should be made strongly typed. `eyre` is not an appropriate error handling solution for libraries. We're slowly removing it so this might be a good time to remove it from block validation api

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:321]: The cloning was introduce by someone other than me in order to solve a bug. This is a immutable reference but it infact has interior mutability. If we don't clone then the wsv is modified and that is incorrect behaviour. We need a local copy we can mutate because we need to validate a series of transactions and how they affect each other.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:268]: Issue created. https://github.com/hyperledger/iroha/issues/3353

File: core/src/sumeragi/main_loop.rs
Comment (SamHSmith) [reply core/src/sumeragi/main_loop.rs:588]: I'm not that convinced. But I wouldn't be against it if you think it's valuable. In that case though I think it's better if you implemented the way you want in a seperate PR. I'm not going to do it justice. Here is the issue I created.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:89]: Pasting my comment from other conversation:
The cloning was introduce by someone other than me in order to solve a bug. This is a immutable reference but it infact has interior mutability. If we don't clone then the wsv is modified and that is incorrect behaviour. We need a local copy we can mutate because we need to validate a series of transactions and how they affect each other.

Comment (SamHSmith) [reply core/src/block.rs:89]: The plan is to replace this with change sets. But that's will wait on general WSV refactoring.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:116]: I don't know. This comment is not mine, I'm simply forwarding it.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:604]: I would like to keep this issue focused on the collapsing of the block types. Therefore I have created this issue concerning eyre, https://github.com/hyperledger/iroha/issues/3353

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:429]: Normally I'm not for future proofing but in this case I think it is important. By having all the methods on VCB we cover for the introduction of other block versions that do things differently. Currently we assume that a v1 block will always be convertible to a v2 block. This is not the case. Imagine just the simple scenario that after a v2 block has been committed we do not want to allow any more v1 blocks. As I've written it we simply switch on the block type and can implement this logic. The previous model were we convert everything into the latest version and use that would not support this case.

Comment (ilchu) [reply core/src/block.rs:424]: I see, that makes quite some sense. Just had a natural urge to completely get rid of code reduplication, but it's not always achievable in real life.

File: core/src/block.rs
Comment (ilchu) [reply core/src/block.rs:89]: Gotcha, just was a bit confusing without some clarifying comment there. I do vaguely remember such conversation happening, though.

**Review by ilchu — APPROVED** [review]

File: core/src/block.rs
Comment (Erigara) [reply core/src/block.rs:89]: Imo we can avoid such confusion in future by accepting `WorldStateView` as parameter, not `&WorldStateView`.

File: core/src/block.rs
Comment (Erigara) [inline core/src/block.rs:435]: In case of `CommittedBlock` we only want to revalidate number of signatures and check transactions hashes.

**Review by Erigara — CHANGES_REQUESTED** [review]
I see that this PR overlapping with #3356 in terms of block refactoring so we need to decide how we gonna merge them.
As i wrote earlier i'm for leveraging type system to prevent accidental bugs and it's what Marin did with introduction of `Candidate<T>`.
So what should we do? @SamHSmith @mversic

File: core/benches/validation.rs
Comment (SamHSmith) [reply core/benches/validation.rs:184]: The refactor broke it and it was not a good benchmark. I could have rewritten it to test how fast we could revalidate the block. But then again the block has only one transaction. It is completely pointless, all it measures is the overhead.

File: core/src/block.rs
Comment (mversic) [reply core/src/block.rs:435]: > There is a bug in validation

in general, I don't we should be structuring our code around an assumption that it's buggy. Maybe this case is different, I won't go into it deeply

Comment (SamHSmith) [reply core/src/block.rs:435]: > in general, I don't think we should be structuring our code around an assumption that it's buggy. Maybe this case is different, I won't go into it deeply

I agree. But I believe this is a case of "don't trust, verify" that we should be upholding as a blockchain project.

File: core/src/sumeragi/message.rs
Comment (Erigara) [inline core/src/sumeragi/message.rs:86]: Why not `VersionedPendingBlock`?

File: core/src/sumeragi/message.rs
Comment (Erigara) [inline core/src/sumeragi/message.rs:138]: If we anyway do full re-validation on block sync there is no point in transferring  `VersionnedCommittedBlock` we can just use `VersionedPendingBlock`.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:435]: @appetrosyan Can we have your opinion on this matter?

File: core/src/block.rs
Comment (Erigara) [inline core/src/block.rs:48]: Maybe `BlockBuilder`?

File: core/src/block.rs
Comment (Erigara) [inline core/src/block.rs:34]: What about just `Block`?

File: core/src/block.rs
Comment (appetrosyan) [reply core/src/block.rs:435]: I agree with both to an extent. 

Right now, we should do the full re-verification, until we conduct a full audit of `sumeragi` our main assumption should be that the code is exceedingly buggy, unless proven otherwise. 

In the future, when we can formally verify that the business logic was executed correctly; which at present doesn't seem possible, but given several audits, and thorough test coverage and fuzzing, should be doable; we should short-circuit double checking.

File: core/src/block.rs
Comment (appetrosyan) [reply core/src/block.rs:34]: `PendingBlock` is much more informative.

File: core/src/sumeragi/message.rs
Comment (appetrosyan) [reply core/src/sumeragi/message.rs:138]: Bad idea. The revalidation could be a precaution, but semantically  the return type is `VersionedCommittedBlock`. Keeping it that way helps with formal verification.

File: core/src/block.rs
Comment (appetrosyan) [reply core/src/block.rs:89]: Rust in general prefers allocations be pushed as far up the call stack as possible. Thus accepting a `wsv` by reference to then clone is considered a bad practice (with caveats).

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:86]: Blocks by consensus are always of the "most recent" version. And instead of promoting all versions to the latest version we should just use the "latest version" in the code. We can make an informed decision about how to handle live upgrades to a network in the future. The current approach gave the impression that this scheme was thought out when I don't think it was. Having a versioned committed block is important though.

I forgot to remove VersionedPendingBlock and have now done so.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs:48]: I will rename it.

**Review by appetrosyan — APPROVED** [review]

File: core/src/sumeragi/message.rs
Comment (Erigara) [reply core/src/sumeragi/message.rs:138]: Yes, untyped, unverified and untrusted and thats why we want to indicate that by type and enforce revalidation When receiving such type through the network.

File: core/src/block.rs
Comment (Erigara) [reply core/src/block.rs:85]: Why this was resolved? there is neither changes nor ticket or reasons why this won't work?
Imo this change is pretty small to be separate ticket.

Comment (SamHSmith) [reply core/src/block.rs:85]: It felt like bikeshedding to me since it will be returned to a pass by reference anyway after we have implemented the new-wsv.

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:138]: In my opinion this would cause again the large amount of types. I think we should avoid turning our programming problems into typesystem problems. But if this was introduced again by another PR I would not be against it.

Comment (mversic) [reply core/src/sumeragi/message.rs:138]: please do not resolve discussions without getting a green light from the OP. 

Also, in this case, it is @Erigara 's opinion as well as mine that you should return the candidate blocks in this PR. Nobody is going to be reverting your changes after you. I will not be going into discussion why it's better to enforce invariants in the type system, it's a common sense. All you should know atm is that majority vote is against you on this topic and we govern ourselves democratically as a team

File: core/src/block.rs
Comment (mversic) [reply core/src/block.rs:85]: whether or not it's bikeshedding, you should not be resolving discussions as you see fit. Discussion can only be closed if given permission by the OP or other. Mutual agreement has to be reached

File: core/src/sumeragi/message.rs
Comment (appetrosyan) [reply core/src/sumeragi/message.rs]: We wouldn't have made it a hard requirement to resolve all conversations before merging if it wasn't necessary. 

Also, the **whole point of code review** is to identify areas where a separate pair of eyes has a different opinion. @Erigara and @mversic have made it clear that they feel strongly about one way of doing things. You don't get to ignore that opinion, you **muat address it**.

File: core/src/block.rs
Comment (Erigara) [inline core/src/block.rs:288]: The same comment about implicit `wsv` clone.

File: core/src/sumeragi/main_loop.rs
Comment (Erigara) [inline core/src/sumeragi/main_loop.rs:602]: This is not genesis transaction so must be

```suggestion
                .revalidate::<false>(
```

File: core/src/sumeragi/main_loop.rs
Comment (Erigara) [inline core/src/sumeragi/main_loop.rs:613]: Here too:

```suggestion
                if let Err(error) = block.revalidate::<false>(
```

File: core/src/sumeragi/main_loop.rs
Comment (Erigara) [inline core/src/sumeragi/main_loop.rs:617]: If we are revalidating against `finalized_wsv` shouldn't we use appropriate `block_height` and `block_hash`?

File: core/src/sumeragi/message.rs
Comment (Erigara) [inline core/src/sumeragi/message.rs:182]: This method is redundant, it's used in only in single place where we can just call `validate_and_extract_block` two times (we can chain them using [`or_else`](https://doc.rust-lang.org/std/result/enum.Result.html#method.or_else)).

File: core/src/sumeragi/message.rs
Comment (Erigara) [inline core/src/sumeragi/message.rs:156]: In my opinion it will be better to enforce type restrictions as early as possible (preferable at system boundary ) and not on the spot.
In our case it will be blocks received as part of `ShareBlocks` message.

File: core/src/sumeragi/message.rs
Comment (Erigara) [inline core/src/sumeragi/message.rs:88]: We were talking about revalidation of `CommittedBlock`, but the same applies for the `PendingBlock`.
We can't trust that we receive validated block and must revalidate it.

**Review by Erigara — CHANGES_REQUESTED** [review]
In general i think it's possible to minimize umber of block states to 4: `PendingBlock`, `Candidate<PendingBlock>`, `CommittedBlock` and `Candidate<ComittedBlock>`.

And with smaller amount we start to lose information about block invariants.

File: core/src/sumeragi/message.rs
Comment (Erigara) [inline core/src/sumeragi/message.rs:163]: Initial idea behind `const generic` was that we know exact places in code where we expect block with certain height (0 or more than zero) and know we fall back to runtime check.

File: core/src/sumeragi/main_loop.rs
Comment (SamHSmith) [reply core/src/sumeragi/main_loop.rs:617]: true!

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:182]: The chaining causes borrow checker problems but I will make your change nonetheless.

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:156]: I don't think that works because block sync does not know if the block is valid or not. Only sumeragi knows because validity is relative to the current state.

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:88]: Is that not what sumeragi is for? ValidatingPeers + ProxyTail all validate the block and the leader does not because it created it? Maybe there's a case for the ObservingPeers but they are supposed to receive all their blocks through block sync. (I know this is currently not the case but I have a ticket for this to do.)

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:163]: I have created this issue, https://github.com/hyperledger/iroha/issues/3372.

File: core/src/sumeragi/message.rs
Comment (Erigara) [reply core/src/sumeragi/message.rs:88]: Yeah, but validating peers and proxy tail receive `BlockCreated` message and we want to make sure that peer don't accidentally forget to validate it.

File: core/src/sumeragi/message.rs
Comment (Erigara) [reply core/src/sumeragi/message.rs:163]: I mean that we still can use `const generics`, not that we should remove them. Because we only expect genesis block inside wait for genesis loop, no?
I don't think that soft fork on genesis block is possible.

File: core/src/sumeragi/main_loop.rs
Comment (Erigara) [inline core/src/sumeragi/main_loop.rs:348]: Hashes are already checked inside `validate_and_extract` block, so we can remove this call.

File: core/src/sumeragi/message.rs
Comment (Erigara) [reply core/src/sumeragi/message.rs:156]: > Only sumeragi knows because validity is relative to the current state.

Yeah and thus it won't be possible to extract `CommittedBlock` inside block sync so we prevent wrong usage of it.

File: core/src/block.rs
Comment (SamHSmith) [reply core/src/block.rs]: This change has been made now.

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:156]: So this conversation can be resolved?

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:163]: The genesis wait loop calls this function because it recieves block sync messages.

Comment (SamHSmith) [reply core/src/sumeragi/message.rs:163]: Should I create a seperate version of the extract function for use by the genesis code only?

File: core/src/sumeragi/main_loop.rs
Comment (SamHSmith) [reply core/src/sumeragi/main_loop.rs:348]: True! And the function can be removed as well.

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:88]: I have now added this.

**Review by Erigara — APPROVED** [review]

**Review by appetrosyan — APPROVED** [review]

**Pull request merged**
